File: tools/sdks/compose/src/main/kotlin/io/thunder/compose/components/presentation/organization/OrganizationSwitcher.kt
package io.thunder.compose.components.presentation.organization

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.text.BasicText
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.*
import androidx.compose.ui.unit.dp
import io.thunder.android.Organization
import io.thunder.compose.LocalThunder
import kotlinx.coroutines.launch

/** Picker to switch the active organization (spec §8.4 Presentation). */
@Composable
fun ThunderOrganizationSwitcher(modifier: Modifier = Modifier) {
    val state = LocalThunder.current
    val i18n = state.i18n
    BaseThunderOrganizationSwitcher(modifier = modifier) { orgs, current, isSwitching, error, switchOrg ->
        Column {
            if (orgs.isEmpty()) BasicText(i18n.resolve("organizationSwitcher.empty"))
            orgs.forEach { org ->
                BasicText(
                    org.name,
                    modifier = Modifier
                        .defaultMinSize(minHeight = 44.dp)
                        .clickable(enabled = !isSwitching) { switchOrg(org) }
                        .semantics {
                            contentDescription = org.name
                            selected = org.id == current?.id
                        },
                )
            }
            error?.let { BasicText(it) }
        }
    }
}

/** Unstyled base variant (spec §8.3). */
@Composable
fun BaseThunderOrganizationSwitcher(
    modifier: Modifier = Modifier,
    content: @Composable (List<Organization>, Organization?, Boolean, String?, (Organization) -> Unit) -> Unit,
) {
    val state = LocalThunder.current
    val scope = rememberCoroutineScope()
    var orgs by remember { mutableStateOf<List<Organization>>(emptyList()) }
    var current by remember { mutableStateOf<Organization?>(null) }
    var isSwitching by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }

    LaunchedEffect(Unit) {
        isSwitching = true
        orgs = runCatching { state.client.getMyOrganizations() }.getOrDefault(emptyList())
        current = runCatching { state.client.getCurrentOrganization() }.getOrNull()
        isSwitching = false
    }

    val switchOrg = { org: Organization ->
        scope.launch {
            isSwitching = true; error = null
            try {
                state.client.switchOrganization(org)
                current = org
                state.refresh()
            } catch (e: Exception) { error = e.message }
            isSwitching = false
        }
    }

    Box(modifier = modifier) { content(orgs, current, isSwitching, error) { switchOrg(it) } }
}

File: tools/sdks/swiftui/Sources/ThunderSwiftUI/components/presentation/organization/OrganizationSwitcher.swift
import SwiftUI
import Thunder

/// Picker to switch the active organization (spec §8.4 Presentation).
public struct ThunderOrganizationSwitcher: View {
    @EnvironmentObject private var i18n: ThunderI18n

    public init() {}

    public var body: some View {
        BaseThunderOrganizationSwitcher { orgs, current, isSwitching, error, switchOrg in
            VStack(alignment: .leading, spacing: 0) {
                if orgs.isEmpty {
                    Text(i18n.resolve("organizationSwitcher.empty"))
                } else {
                    ForEach(orgs, id: \.id) { org in
                        Button {
                            if !isSwitching { Task { await switchOrg(org) } }
                        } label: {
                            HStack {
                                Text(org.name)
                                if org.id == current?.id {
                                    Spacer()
                                    Image(systemName: "checkmark")
                                }
                            }
                        }
                        .accessibilityLabel(org.name)
                        .accessibilityAddTraits(org.id == current?.id ? .isSelected : [])
                        .frame(minWidth: 44, minHeight: 44)
                        .disabled(isSwitching)
                    }
                }
                if let error { Text(error).foregroundColor(.red) }
            }
        }
    }
}

/// Unstyled base variant (spec §8.3).
public struct BaseThunderOrganizationSwitcher<Content: View>: View {
    @EnvironmentObject private var state: ThunderState
    public let content: ([Organization], Organization?, Bool, String?, (Organization) async -> Void) -> Content

    @State private var orgs: [Organization] = []
    @State private var current: Organization?
    @State private var isSwitching = false
    @State private var error: String?

    public init(
        @ViewBuilder content: @escaping ([Organization], Organization?, Bool, String?, (Organization) async -> Void) -> Content
    ) {
        self.content = content
    }

    public var body: some View {
        content(orgs, current, isSwitching, error, switchOrg)
            .task { await load() }
    }

    private func load() async {
        isSwitching = true
        async let myOrgs = state.client.getMyOrganizations()
        async let cur = state.client.getCurrentOrganization()
        orgs = (try? await myOrgs) ?? []
        current = try? await cur
        isSwitching = false
    }

    private func switchOrg(_ org: Organization) async {
        isSwitching = true; error = nil
        do {
            _ = try await state.client.switchOrganization(org)
            current = org
            await state.refresh()
        } catch { self.error = error.localizedDescription }
        isSwitching = false
    }
}
